dotnet/Razorvine.Pyrolite/Pyrolite/Pickle/ArrayReader.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace Razorvine.Pickle
+{
+    internal struct ArrayReader : IInputReader
+    {
+        private readonly byte[] input;
+        private int position;
+
+        public ArrayReader(byte[] bytes)
+        {
+            input = bytes;
+            position = 0;
+        }
+
+        public byte ReadByte()
+        {
+            return input[position++];
+        }
+
+        public ReadOnlySpan<byte> ReadBytes(int bytesCount)
+        {
+            var result = new ReadOnlySpan<byte>(input, position, bytesCount);
+            position += bytesCount;
+            return result;
+        }
+
+        public string ReadLine(bool includeLF = false)
+        {
+            return PickleUtils.rawStringFromBytes(ReadLineBytes(includeLF));
+        }
+
+        public ReadOnlySpan<byte> ReadLineBytes(bool includeLF = false)
+        {
+            var result = ReadBytes(GetLineEndIndex(includeLF));
+            if (!includeLF)
+                Skip(1);
+            return result;
+        }
+
+        public void Skip(int bytesCount)
+        {
+            position += bytesCount;
+        }
+
+        private int GetLineEndIndex(bool includeLF = false)
+        {
+            var bytes = new ReadOnlySpan<byte>(input, position, input.Length - position);
```

Comment:
(nit) could simplify this to:

```C#
var bytes = input.AsSpan(position;
```

Reply:
good suggestion, unfortunately, there is no `AsSpan` that would create a **ReadOnly**Span from array.

```cs
public static class MemoryExtensions
{
    public static ReadOnlySpan<char> AsSpan(this string text);
    public static Span<T> AsSpan<T>(this T[] array, int start, int length);
    public static ReadOnlySpan<char> AsSpan(this string text, int start);
    public static ReadOnlySpan<char> AsSpan(this string text, int start, int length);
    public static Span<T> AsSpan<T>(this ArraySegment<T> segment, int start, int length);
    public static Span<T> AsSpan<T>(this ArraySegment<T> segment, int start);
    public static Span<T> AsSpan<T>(this ArraySegment<T> segment);
    public static Span<T> AsSpan<T>(this T[] array, int start);
    public static Span<T> AsSpan<T>(this T[] array);
```

/cc @AhsonKhan

I could do:

```cs
ReadOnlySpan<byte> bytes = input.AsSpan(position);
```

but it would always include implicit cast from Span to ReadOnlySpan which would have (minor but still) effect on perf

Reply:
I don't think it should have noticeable affect on perf. This cast should effectively be "free". It's a 2-field struct copy: https://source.dot.net/#System.Private.CoreLib/shared/System/Span.Fast.cs,302

Reply:
I know that it's a cheap operation, but I want to avoid every possible overhead.